src/libraries/System.Linq/src/System/Linq/Select.SizeOpt.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private sealed class SizeOptIListSelectIterator<TSource, TResult>(IList<TSource> _source, Func<TSource, TResult> _selector)
+            : Iterator<TResult>
+        {
+            public override int GetCount(bool onlyIfCheap) => _source.Count;
```

Comment:
This looks like it's going to have the same issue as when implementing IList; Count() will invoke this and the _selector won't be invoked. It instead needs to do something like this:
https://github.com/dotnet/runtime/blob/0720de0a82725160e1ddf40f1db4fe0b9ea186c8/src/libraries/System.Linq/src/System/Linq/Select.SpeedOpt.cs#L428-L444

Reply:
I'll double check. I did run our entire normal suite locally with SizeOpt enabled.